### PR TITLE
`WizardAppWidget`: add `open_first_step` parameter, True by default

### LINKED ISSUE #793: WizardAppWidget starts fully closed in ipywidgets=8
WizardAppWidgets uses `ipywidgets.Accordion` to control which step is opened. 
`Accordion` (and other similar widgets) [use `selected_index` trait](https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20List.html#accordion-tab-and-stack-use-selected-index-not-value) to determine which step is opened. Value of None closes all steps, value 0 opens the first step.

The default `selected_index` value changed in ipywidgets 8 from 0 to None so now all steps are closed when the app is first loaded. (Note that QeApp already had this even before, which is why this was not noticed).

I propose that the we should add a parameter that determines this behaviour, e.g. `open_first_step`.
To preserve backwards compatibility, it should be True by default, but if people feel strongly that this default would be wrong, we can also change it.

One point in favour of not changing it is that this issue was causing a bug in AtmoSpec (which is how I noticed), which the nglview viewer didn't work, because apparently the viewer is not properly initialized if the containing ipywidgets is not rendered (opened). QeApp has a horrible workaround for this issue

https://github.com/aiidalab/aiidalab-qe/blob/a4e4280c37aaad1dba52e3fa23c9602cf7c341a3/src/aiidalab_qe/app/structure/step.py#L191

```python
    def _post_render(self):
        # After rendering the widget, nglview needs to be resized
        # to properly display the structure
        self.manager.viewer._viewer._set_size("100%", "300px")
```

I haven't yet figured out how to use this workaround in my app since I don't have the MVC architecture. 
In any case, if we don't restore the previous default behaviour, other application may run into this issue as well if they have a Structure widget in the first step.

### PR DESCRIPTION
**Pull request by danielhollas**
In ipywidgets v8, the default `selected_index` changed from 0 to None,
which means that the WizardAppWidget now starts fully closed.

We want the app developers to have control over this so we add a new parameter `open_first_step` and we have it True by default to be backwards compatible.

In a second commit I added a bit of typing to `wizard.py` so best reviewed commit by commit.

Closes #793. See more details in this issue. 

(happy to bikeshed the parameter name)

### CODE REVIEW

**Review by edan-bainglass — APPROVED** [review]
LGTM! TBD if we want to replace this with the MVC-based one in QE, i.e., move the aiidalab_qe/common/wizard package to AWB.

**Comment by edan-bainglass** [comment]
@danielhollas I approved this rather quickly, as it is fine in principle, though can't one just set the `selected_index` to their chosen default step? Setting it to 0 is effectively what `open_first_step` does, no?

**Comment by danielhollas** [comment]
> though can't one just set the selected_index to their chosen default step? Setting it to 0 is effectively what open_first_step does, no?

Sure, but then you'd be accessing implementation details of the class (via self.accordion)

**Comment by edan-bainglass** [comment]
> Sure, but then you'd be accessing implementation details of the class (via self.accordion)

```python
class Wizard(ipw.Accordion):
    def __init__(
        self,
        steps: dict[str, ipw.VBox],
        selected_index: int | None = None,  # or 0, TBD
    ):
        super().__init__(children=list(steps.values()))

        for i, title in enumerate(steps.keys()):
            self.set_title(i, title)

        self.selected_index = selected_index


STEPS = {
    "1. Input": ipw.VBox([ipw.HTML("<h3>Step 1: Input</h3>")]),
    "2. Workflow": ipw.VBox([ipw.HTML("<h3>Step 2: Workflow</h3>")]),
    "3. Output": ipw.VBox([ipw.HTML("<h3>Step 3: Output</h3>")]),
}

wizard = Wizard(steps=STEPS, selected_index=None)
```
We can call our `selected_index` whatever, say, `default_step`. My point is to keep it general, in case a developer wishes to open their wizard in step 2 by default (for some reason). With `open_first_step`, we hardcode "first" 🙂

**Comment by danielhollas** [comment]
Oh, I see that make sense, I'll update the PR.

**Review by edan-bainglass — APPROVED** [review]
LGTM! Thanks @danielhollas. I'm only wondering if `default_step` is more user friendly, though one may argue that `selected_index` is more consistent 🤷🏻‍♂️ @yakutovicha thoughts?

**Comment by yakutovicha** [comment]
> LGTM! Thanks @danielhollas. I'm only wondering if `default_step` is more user friendly, though one may argue that `selected_index` is more consistent 🤷🏻‍♂️ @yakutovicha thoughts?

If unclear, I would choose whatever ipywidgets have chosen, at least to be in sync. In that case, `selected_index`.

File: tests/test_wizard.py [hunk 1]
```diff
@@ -4,6 +4,54 @@
 from aiidalab_widgets_base import WizardAppWidget, WizardAppWidgetStep
 
 
+class Step1(ipw.HBox, WizardAppWidgetStep):
```
Comment (yakutovicha) [inline tests/test_wizard.py:9]: It is literally a copy of the classes defined in the test below - maybe you should delete the ones in the `test_wizard_app_widget()`?

File: tests/test_wizard.py [hunk 2]
```diff
@@ -118,3 +171,20 @@ def _observe_config(self, change):
...
+    widget = WizardAppWidget(
+        steps=[
+            ("Setup", s1),
+            ("View results", s2),
+        ],
+        selected_index=None,
+    )
+
+    # All steps should be closed initially
+    # First step should be opened by default
```
Comment (yakutovicha) [inline tests/test_wizard.py]: leftover?
```suggestion
```

File: tests/test_wizard.py [hunk 3]
```diff
@@ -118,3 +171,20 @@ def _observe_config(self, change):
...
+    assert widget.accordion.selected_index is None
```
Comment (yakutovicha) [inline tests/test_wizard.py:145]: Please also test that the Navigation buttons are correctly disabled when the selected index is `None`.

```suggestion
    assert widget.accordion.selected_index is None
    # With no step selected, all header buttons should be disabled.
    assert widget.back_button.disabled is True
    assert widget.next_button.disabled is True
    assert widget.reset_button.disabled is True
```

File: tests/test_wizard.py [hunk 4]
```diff
@@ -118,3 +171,20 @@ def _observe_config(self, change):
     assert widget.accordion.selected_index == 1
     assert widget.next_button.disabled is True
     assert widget.back_button.disabled is False
+
+
+def test_closed_first_step():
```
Comment (yakutovicha) [inline tests/test_wizard.py:132]: Since this is a new possibility, please also test when the index at initialization is 1:

```python
def test_selected_index_at_construction():
    s1 = Step1()
    s2 = Step2()

    widget = WizardAppWidget(
        steps=[
            ("Setup", s1),
            ("View results", s2),
        ],
        selected_index=1,
    )

    # The widget should open directly on the requested step ...
    assert widget.accordion.selected_index == 1
    assert widget.selected_index == 1
    # ... and buttons should reflect that step's state immediately,
    # without requiring any further interaction.
    assert widget.back_button.disabled is False # this is NOT the case atm
    assert widget.next_button.disabled is True  # s2 is still in State.INIT
```

And, more generally, what do we mean when the index is set to 1? How `back_button`/`next_button` should be displayed? Especially when step 0 isn't successful, should `back_button` be enabled to invite the user to go back? If that is the case, then the current implementation is not doing that currently.

File: aiidalab_widgets_base/wizard.py [hunk 5]
```diff
@@ -88,19 +91,23 @@ class WizardAppWidget(ipw.VBox):
 
     selected_index = tl.Int(allow_none=True)
 
-    def __init__(self, steps, show_header=True, **kwargs):
+    def __init__(
+        self,
+        steps: Sequence[tuple[str, WizardAppWidgetStep]],
+        show_header: bool = True,
+        selected_index: int | None = 0,
```
Comment (yakutovicha) [inline aiidalab_widgets_base/wizard.py]: Please add validation when the `selected_index` is out of range. Also, make sure to have a test for that.
Something like:
```python
        if selected_index is not None and not 0 <= selected_index < len(steps):
            raise ValueError(
                f"selected_index must be None or in range [0, {len(steps) - 1}], "
                f"got {selected_index}."
            )
```

File: aiidalab_widgets_base/wizard.py [hunk 6]
```diff
@@ -5,7 +5,10 @@
     * Carl Simon Adorf <simon.adorf@epfl.ch>
 """
 
+from __future__ import annotations
```
Comment (yakutovicha) [inline aiidalab_widgets_base/wizard.py:8]: Just flagging that this is a drive-by fix and, strictly speaking, not part of the issue.

**Review by yakutovicha — CHANGES_REQUESTED** [review]
Thanks very much @danielhollas, I have a few minor requests.

Also, a side remark about type annotations. I understand the desire to clean up other things while doing the work, but let's stay disciplined about keeping that to separate PRs. I see two main reasons for doing that:

- It keeps PRs focused and goal-oriented, so later it is easier to find what was done by looking at the title/description.
- If a change needs to be reverted, it reverts other unexpected things (type annotations in that case).

This is not a blocker, but if you have some time, please split them up. Otherwise, go ahead, but remember for the future 🙏 .

File: aiidalab_widgets_base/wizard.py (hunk 5 above)
Comment (danielhollas) [reply aiidalab_widgets_base/wizard.py]: I thought about adding explicit validation at first, but we would essentially be duplicating the validation that Accordion already does. 
So I don't think it's worth it. 

I should certainly add a test (if I hadn't already).

File: tests/test_wizard.py (hunk 1 above)
Comment (danielhollas) [reply tests/test_wizard.py:9]: Oh, this was certainly my intention, I just forgot to delete them, thanks for catching this!

**Comment by danielhollas** [comment]
> Also, a side remark about type annotations. I understand the desire to clean up other things while doing the work, but let's stay disciplined about keeping that to separate PRs.

Agreed, I've opened #796 and rebased this one on top of it. Marking as draft while I work on the review, thanks!

File: aiidalab_widgets_base/wizard.py (hunk 5 above)
Comment (danielhollas) [reply aiidalab_widgets_base/wizard.py]: Added tests in 72d4184

File: tests/test_wizard.py (hunk 3 above)
Comment (danielhollas) [reply tests/test_wizard.py:145]: Thanks, added in 8c7c34f

File: tests/test_wizard.py (hunk 2 above)
Comment (danielhollas) [reply tests/test_wizard.py]: Yep

File: tests/test_wizard.py (hunk 4 above)
Comment (danielhollas) [reply tests/test_wizard.py:132]: I added your proposed test in 9731edd and it passes so I think the buttons are behaving correctly here, because the existing logic already handles the case where Step 2 is opened while Step 1 is not finished (you can do that manually by clicking, this is not new in this PR).

> And, more generally, what do we mean when the index is set to 1?

Well, this is why I originally only had a boolean `opened_first_step`, because I can't imagine the case where setting `selected_index` to anything other than 0 or None is useful. WDYT? We might want to revisit this.

**Comment by danielhollas** [comment]
https://github.com/aiidalab/aiidalab-widgets-base/pull/792#issuecomment-5289938045

> We can call our selected_index whatever, say, default_step. My point is to keep it general, in case a developer wishes to open their wizard in step 2 by default (for some reason). With open_first_step, we hardcode "first" 🙂

I am still not sure about this. "Keeping it general" feels like it results in a somewhat confusing API, since I can't imagine how the fully generality is useful at all (i.e. why would you *ever* want your app to be opened at second step???

File: aiidalab_widgets_base/wizard.py (hunk 5 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/wizard.py]: > I thought about adding explicit validation at first, but we would essentially be duplicating the validation that Accordion already does.

It would give a little bit clearer error message that belongs to the wizard, not the accordion underneath. But either way is fine.

**Comment by yakutovicha** [comment]
> [#792 (comment)](https://github.com/aiidalab/aiidalab-widgets-base/pull/792#issuecomment-5289938045)
> 
> > We can call our selected_index whatever, say, default_step. My point is to keep it general, in case a developer wishes to open their wizard in step 2 by default (for some reason). With open_first_step, we hardcode "first" 🙂
> 
> I am still not sure about this. "Keeping it general" feels like it results in a somewhat confusing API, since I can't imagine how the fully generality is useful at all (i.e. why would you _ever_ want your app to be opened at second step???

I also cannot imagine the use case for keeping it general, but that doesn't mean there isn't. Ultimately, I don't see severe consequences of any choice, tbh. So my opinion is as much of a value as tossing a coin. Claude is also unsure :).

**Review by yakutovicha — APPROVED** [review]

**Comment by edan-bainglass** [comment]
@danielhollas I got a ping for review, but this is closed?

**Comment by danielhollas** [comment]
Stupid github closed the PR after I merged #796. I will need to reopen, but the review request was mostly for the discussion regarding the discussion about `open_first_step` versus `selected_index`. Please see the last two comments and feel free to reply here.

**Comment by danielhollas** [comment]
Or maybe I can reopen?

EDIT: Apparently not

**Comment by danielhollas** [comment]
Let's try re-opening again

**Comment by edan-bainglass** [comment]
> Stupid github closed the PR after I merged #796. I will need to reopen, but the review request was mostly for the discussion regarding the discussion about `open_first_step` versus `selected_index`. Please see the last two comments and feel free to reply here.

I'm still in favor of not hard coding `first_step` but also against the vague `selected_index`. I had suggested `default_step` as an alternative.

**Comment by danielhollas** [comment]
Haha, looks like we're not converging. :sweat_smile: 
(sorry for the bikeshedding here, but since this is a public API it's worth to discuss a bit).

> We can call our selected_index whatever, say, default_step. My point is to keep it general, in case a developer wishes to open their wizard in step 2 by default (for some reason). With open_first_step, we hardcode "first" 🙂

> I'm still in favor of not hard coding first_step but also against the vague selected_index. I had suggested default_step as an alternative.

@edan-bainglass you seem to have the firmest opinion here, but could you expand your reasoning a little bit?

I understand the appeal of generality, but in this case this feels like a case of [premature abstraction](https://arendjr.nl/blog/2024/07/post-architecture-premature-abstraction-is-the-root-of-all-evil/#:~:text=I%20would%20adapt%20one,help%20maintainability.). 

I was originally like Sasha roughly fifty-fifty between the options, but now I am leaning more to my original boolean solution, because

1. It feels more obvious and easier to explain
2. You don't have to worry about validation. 
3. `default_step` feels confusing, and it's not even clear what type it should be. 
4. `default_step=None` means all steps are closed at the beginning, but that is not obvious. 
5. Seriously, if you think your `default_step` should be anything other than the first step, that feels like a design issue with the application, and goes against the concept of WizardStep widget imo.


I'll let this sit over the weekend, and unless convinced or otherwise stopped I am going to make an executive decision on Tuesday, as the author of the PR. :-) 

EDIT: Marking as draft to make it clear we have a decision to make

**Comment by edan-bainglass** [comment]
The first step of my hypothetical wizard is optional (carries some defaults that are applicable most of the time). It still represents the first step (or rather, a pre-step), as it affects the next. But I prefer to start the app in the second step, with a friendly message that the user may visit the first step to change defaults if they wish. But by default, I'd want the `default_step` to be 2 (equal to `selected_index = 1`. If `default_step` is then a step index >= 1, with 0 or None being all are closed (that last part is not ideal). We could also just use `selected_index` and match the underlying accordion API. My point is that if we don't generalize, then tomorrow someone comes with that application, as it tends to happen.

Anyhow, let's proceed democratically. If @danielhollas and @yakutovicha believe that this is not necessary, carry on. If and when this use case comes around, we can create a wrapper `Wizard` to address it, and we'll have a concrete answer for the next major version 🙂 I'm fine with it.

**Comment by danielhollas** [comment]
@yakutovicha given the discussion above, are you still fifty-fifty on the two proposed solutions?

**Comment by yakutovicha** [comment]
> @yakutovicha given the discussion above, are you still fifty-fifty on the two proposed solutions?

Yes, mostly because I don't see a severe negative consequence of any decision. I spent some time trying to find a definitive factor that would make one decision definitely preferable - I didn't find any.

What I do have an opinion about is "a bad decision is better than no decision", simply because a bad decision can be fixed. 

@danielhollas, let's be pragmatic here; we've already spent enough time. Since this particular change is not a high-stakes call, please go ahead with your `open_first_step: bool` (or any other name you want), and if a new use case comes out, we can generalize that in the future.

**Comment by danielhollas** [comment]
Alright, thank you both for the discussion.

> @danielhollas, let's be pragmatic here; we've already spent enough time. Since this particular change is not a high-stakes call, please go ahead with your open_first_step: bool (or any other name you want), and if a new use case comes out, we can generalize that in the future.

Yep, in the end I went with `open_first_step` as a simpler solution. If the need arises in the future, I will first eat my hat, and then implement `selected_index`. :-)
